kernel: add bootargs support

fkokosinski opened the pull request:
This PR adds support for receiving bootargs from the loader and passing them to the main after splitting them into arguments. It includes implementation for two loaders (multiboot and zefi).

To support additional loaders the `const char *get_bootargs()` has to be implemented for them. After enabling it and one of the supported loaders the main function in the form `int main(int argc, char **argv)` can be used.

This PR also includes a sample with tests.

jmichalski-ant committed:
zefi: add bootargs support

Add bootargs support to zefi. This implements
get_bootargs() when both efi and bootargs are
selected in config.

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

jmichalski-ant committed:
kernel: add bootargs support

Add support for passing args to main(). The
content of bootargs is taken from get_bootargs()
which should be implemented for each loader and
then its split into args and passed to main.

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

jmichalski-ant committed:
arch/x86: multiboot: add bootargs support

Add bootargs support for multiboot. This
implements get_bootargs() when multiboot and
bootargs are selected in config.

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

jmichalski-ant committed:
samples: add bootargs sample

Add bootargs sample

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

andyross commented on arch/x86/core/efi.c:
```diff
@@ -169,3 +175,44 @@ int arch_printk_char_out(int c)
+
+	/* efi uses ucs-2, we have to convert it to utf-8 */
+	iconv_t ucs2_to_u8 = iconv_open("UTF-8", "UCS-2LE");
+	size_t iconv_ret = iconv(ucs2_to_u8, &in, &inb, &out, &outb);
```
Hm... IIRC minimal libc doesn't even have iconv, which seems like it would be a non-starter here.

Do you really need to do this, anyway?  It's a kernel command line, not an i18n-sensitive context.  IIRC Linux demands that the command line be 7-bit ASCII (or at least literally nothing assumes otherwise), can't we do the same and just implement this with a range test and cast to a char?

fkokosinski replied:
removed

andyross commented on arch/x86/core/multiboot.c:
```diff
@@ -11,6 +11,43 @@
+/*
+ * We have to map cmdline from multiboot, but we have to know its size
+ * to do it. Its not provided as a separate field, so we have to calculate
+ * it here by mapping it part by part.
```
Similar comments, though I'm much less an expert on multiboot.   This seems way too complicated, trying to extract the string from "foreign" unmapped memory after OS entry.   Multi boot enters the OS in 32 bit mode with paging disabled, doesn't it?  Wouldn't it make much more sense to copy the string up in the stub than it does here?

andyross commented on include/zephyr/arch/x86/efi.h:
```diff
@@ -12,10 +12,20 @@
+	/*
+	 * bootargs fields can't be placed here conditionally based on CONFIG_BOOTARGS
+	 * because this struct is also used in arch/x86/zefi/zefi.c where config defines
+	 * are unavailable
```
I thought we had autoconf.h included in zefi.c already, no?  If not, there's no reason that can't be made to work.  Seems silly to go through the trouble of writing this comment instead of fixing the problem.

andyross commented on include/zephyr/arch/x86/multiboot.h:
```diff
@@ -105,9 +107,10 @@ struct multiboot_mmap {
+#define MULTIBOOT_INFO_FLAGS_MEM		(1 << 0)	/* mem_* valid */
+#define MULTIBOOT_INFO_FLAGS_CMDLINE		(1 << 2)	/* cmdline* valid */
+#define MULTIBOOT_INFO_FLAGS_MMAP		(1 << 6)	/* mmap_* valid */
+#define MULTIBOOT_INFO_FLAGS_FB			(1 << 12)	/* fb_* valid */
```
If you're going to muck with the formatting here, might as well port it to use BIT() instead of the hand-cooked shift.

fkokosinski replied:
using the `BIT` macro now

andyross commented on kernel/init.c:
```diff
@@ -403,6 +406,84 @@ static inline int z_vrfy_device_init(const struct device *dev)
+{
+	const char *bootargs = get_bootargs();
+	static char *argv_buf[CONFIG_BOOTARGS_ARGV_BUFFER_SIZE] = { NULL };
+	static char argv_strings_buf[CONFIG_BOOTARGS_ARGS_BUFFER_SIZE];
```
Nitpick, but with only a little extra code you can put the argv pointers at the end of the buffer while you copy from the start, and avoid having to configure the two separately.

fkokosinski replied:
I believe this is fixed as well

andyross reviewed (commented):
Seems nifty.  Some design notes.  Nothing fatal really, but the MMU contortions required to suck the arguments out of the primordial/inaccessible memory (vs. just copying them in the environment where they can naturally be used) kinda bug me.

jmichalski-ant committed:
zefi: include autoconf.h

Zefi was only including include/zephyr but not the generated include
directory that contains autoconf.h. This commit fixes that.

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

cfriedt commented on kernel/init.c:
```diff
@@ -37,6 +37,9 @@
 #include <zephyr/logging/log.h>
 #include <zephyr/pm/device_runtime.h>
 #include <zephyr/internal/syscall_handler.h>
+#ifdef CONFIG_BOOTARGS
+#include <ctype.h>
+#endif /* CONFIG_BOOTARGS*/
```
No need for conditional inclusion of this header.
```suggestion
#include <ctype.h>
```

fkokosinski replied:
The guarding ifdef was removed and this include was moved before Zephyr’s includes. At the same time other standard includes were moved too, to keep all standard includes in front of Zephyr ones

cfriedt commented on kernel/Kconfig:
```diff
@@ -1006,6 +1006,20 @@ config STATIC_INIT_GNU
 	  Note the option CMAKE_LINKER_GENERATOR does not yet support this feature
 	  or CPP.
 
+config BOOTARGS
+	bool "Support bootargs"
```
Technically speaking, bootargs can always be enabled. In the static case, just default the args to `""` and in the dynamic case, just default the size to 1 (and ensure that the storage is static).

fkokosinski replied:
> Technically speaking, bootargs can always be enabled

That's true, however:

If we keep the bootargs always enabled, `main(int, char **)` will be called unconditionally. Even if it will be linked with `main(void)` later on. White this seems to work just fine (at least on x86, for which it was tested), I fear that this will be prone to randomly break, especially when LTO is enabled.

Plus, the bootargs feature bring some new code that the user might not want to always include. So I'd say - let's leave that choice to the user.

cfriedt reviewed (commented):
Personally, I think it would be better to encode bootargs (aka kernel cmdline) as a string via Kconfig.

However, I realize that there is often a need for dynamic bootargs.

So a good middle ground might be to have a choice; static vs dynamic. If dynamic, then encode the location with Devicetree as a `zephyr,chosen`. This will be usable by both bootloaders such as mcuboot but also by the app. Additionally, the host in a PCI environment (although in the latter, there may need to be a mailbox or something to poll set up).

cfriedt commented on kernel/Kconfig:
```diff
@@ -1006,6 +1006,20 @@ config STATIC_INIT_GNU
 
+config BOOTARGS
+	bool "Support bootargs"
+	default n
```
This line should be removed, since the default for boolean options is already n

fkokosinski replied:
Done

fkokosinski commented on arch/x86/core/efi.c:
```diff
@@ -169,3 +175,44 @@ int arch_printk_char_out(int c)
+	char *efi_bootargs_virt = NULL;
+
+	k_mem_map_phys_bare((uint8_t **)&efi_bootargs_virt, (uintptr_t)efi->efi_bootargs,
+			    efi->efi_bootargs_size, 0);
```
bootargs are now passed in a static buffer

fkokosinski commented on samples/basic/bootargs/README.rst:
```diff
@@ -0,0 +1,97 @@
+.. zephyr:code-sample:: bootargs
```
Moved

fkokosinski commented:
> If dynamic, then encode the location with Devicetree as a zephyr,chosen

Currently, the chosen node in Zephyr assumes that only phandles are assigned in this node. This means that we can’t simply assign a string like in Linux, and the `DT_HAS_CHOSEN` macro won’t work in our case because of that. Unless you had something else in mind that I’ve missed?

We’ve added the static/dynamic differentiation, but left the dynamic implementation as it originally was - just defining arguments for QEMU.

fkokosinski commented:
@kartben @andyross @cfriedt Could you PTAL? Thanks!

cfriedt reviewed (commented):
For x86 / EFI this is probably great, and I am no expert there by any means. I'm thinking more of a generic / arch independent way to specify a kernel cmdline, which is maybe out of scope here.

Since we're talking about C entry points, it might help to get feedback from @stephanosio and @keith-packard.

keith-packard commented:
> Since we're talking about C entry points, it might help to get feedback from @stephanosio and @keith-packard.

Looks fine to me. I did go check the C and POSIX specs to see if the `int main (int argc, char **argv, char **env)` form was mentioned anywhere, but it wasn't. Ofc, we'll need to update the Zephyr docs mentioning the new alternative form for `main`.

jmichalski-ant committed:
docs: add information about main(int, char **)

Add information about main(int, argc **) in places where the documentation
stated that only parameterless main could be used

Signed-off-by: Jakub Michalski <jmichalski@internships.antmicro.com>
Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
> Looks fine to me. I did go check the C and POSIX specs to see if the `int main (int argc, char **argv, char **env)` form was mentioned anywhere, but it wasn't. Ofc, we'll need to update the Zephyr docs mentioning the new alternative form for `main`.

Hi @keith-packard, thanks for your comment!

We've updated the Zephyr docs so that the documentation mentions the alternative form for `main`.

fkokosinski commented:
Hi @andyross, @peter-mitsis - pinging you guys since you're assigned here. Is there anything else that I'm missing that blocks us from moving forward with this PR?

andyross reviewed (approved):
Looks like my gripes were addressed

keith-packard reviewed (approved):
Thanks for adding doc updates.
